### ISSUE TITLE
Honeywell Vista system state not properly reflecting arm night mode 

### ISSUE THREAD

**Issue by blfriedman** [ufodone/envisalink_new#146]
I have configured my Envisalink 4 / Honeywell Vista-20p "Arm Night Mode" for night-stay operation

![image](https://github.com/user-attachments/assets/bca311b0-7b77-4ab9-9ae1-bb62b992d9bb)

I arm with the alarm control panel (physical) with the "33" code.  The physical panel reflects the system is in "Night" mode.

I set up an automation on the state of "Honeywell Alarm System" when set to "Night" mode, but I can see from the log that it is only showing that it is in "Armed home" rather than "Armed Night".  I can see from the UI that it has a night armed capability.

![image](https://github.com/user-attachments/assets/bab27849-bf2f-4310-8e05-4f3f4c50fd80)

Originally, I used the old envisalink integration and had the settings in configuration.yaml, but after migration, I removed them from configuration.yaml as advised.

My hardware
![image](https://github.com/user-attachments/assets/bff3038f-0c9e-4e16-9690-fc2872a436bc)

You know, I could swear that after I migrated to envisalink_new initially many months ago, that night mode worked properly.  I can't say when this broke, unfortunately.

Thanks in advance for any help with this.

**Comment by ufodone** [comment]
The code is supposed to be setting the night mode state should be set if the alarm is indicating that it is armed with no entry delay.  Would that be true for your scenario?  It doesn't look like anything related to that has changed (intentionally) for a very long time.  

If you could send some debug logs going through the arm process I could take a look to see what's going on.

**Comment by blfriedman** [comment]
> The code is supposed to be setting the night mode state should be set if the alarm is indicating that it is armed with no entry delay. Would that be true for your scenario? It doesn't look like anything related to that has changed (intentionally) for a very long time.

First, let me say thanks for creating this integration.  Also, while I think I had this working at one point, but unfortunately, my memory might be faulty.  I do know that with envisalink_new I could arm the system through HA in night mode; something I could not do with the old integration.  I usually do it manually at the alarm keypad anyway.  So, it is possible that the only improvement I saw with envisalink_new was the ability to arm night mode.  However, I created automations specifically for night mode at that time, and I thought they worked.

I expect that if I use the "night-stay" mode to arm - whether with the integration or manually at the alarm keypad, that the  integration alarm state will reflect "night" mode.  The Honeywell 6150 fixed english keypad displays "night" when armed in night mode.

The setting of night mode in the integration based solely on no entry delay (i.e. instant arming mode) seems wrong.  Especially given that there is a configuration item in the integration to select whether arming in night mode is instant or night-stay, and if configured for night-stay, which I am using, it does not use the same configuration item to determine that night mode has been set.

> 
> If you could send some debug logs going through the arm process I could take a look to see what's going on.

I have attached a debug log around arming from last night - after the countdown completes to entering night-stay after arming night-stay at the alarm keypad -

`2024-10-21 23:48:13.579 DEBUG (MainThread) [custom_components.envisalink_new.pyenvisalink.honeywell_client] {"partition_state": "N/A", "alpha": "***NIGHT-STAY***                ", "ac_present": true, "beep": "unknown", "armed_bypass": false, "entry_delay": false, "exit_delay": false, "last_armed_by_user": 2, "last_disarmed_by_user": 2, "ready": false, "bat_trouble": false, "trouble": false, "fire": false, "panic": false, "alarm": false, "alarm_fire_zone": false, "alarm_in_memory": false, "armed_away": false, "armed_stay": true, "armed_zero_entry_delay": false, "chime": true}`

[error_log.txt](https://github.com/user-attachments/files/17478001/error_log.txt)

**Comment by ufodone** [comment]
I think the problem is that the panel does not provide any way to determine whether it has been armed with night-stay mode.  It has flags indicating away and stay as well as zero-entry delay.  So this is probably why the original author was using stay+zero-entry-delay to indicate night-stay mode.  It's true that the keypad may display the NIGHT-STAY text but this integration has generally stayed away from trying to parse that the text strings out as it's unclear whether these strings are consistent across all panels.

It might be possible to track the state within the integration itself but the downside of that is that a restart of HA could result in the state being wrong when it started back up if the alarm was armed/disarmed and changed state while HA was down.  It is definitely an edge case which may or may not matter depending on what you (or others) might be using the state for via automations, etc.

**Comment by blfriedman** [comment]
Thank you for your response.  I looked up the Envisalink Vista TPI programmer's document v1.03 / 2/10/2017.  I don't see any reference there to night-stay mode, as you indicate.

I found discussion from the original envisalink integration author about instant mode stay, but as I mentioned, this is flawed because those are two different modes with different functionalities.

I did some research on the evolution of the ECP bus protocol, but did not have access to the original Circuit Cellar article on reverse engineering the ECP bus - nor could I locate an API/Protocol document.  I did however find this project:
[((https://github.com/search?q=repo%3ADilbert66%2Fesphome-vistaECP%20BIT_MASK_BYTE1_NIGHT&type=code))]

The esphome-vistaECP project has the capability to get/set night mode on the Vista ECP.  I found the reference (see link) to the F7 message bitfield containing the night mode indicator.  So apparently there is an F7 status message that does contain the night mode indicator for night-stay mode.  I'm not sure if Envisalink allows access to this however.

**Comment by ufodone** [comment]
I suspect the EVL devices themselves would use this same interface but this integration only has access to what the TPI exposes.  But I think that due to hardware limitations (largely memory available for the firmware I believe) there is not much room for additional changes to the EVL firmware so the likelihood of them adding additional support is probably very low.

The only other thing I can think of might be the CID codes that the device sends.  I don't really know much about them but if there was a code that indicated night mode I might be able to use that.  But from the logs that you sent, the only CID code I see (via the %03 command) is 441 which apparently is "STAY/INSTANT" and I assume doesn't provide the needed additional context to differentiate between instant and night mode.

**Comment by blfriedman** [comment]
Thank you again.  I am going to follow up directly with EyezOn about the TPI interface for the Ademco interface.  The latest version I have seen documented is 1.03.

The Envisalink TPI states that the CID code is a "ContactID code" and that they are widely available, and they decline to include them in the TPI document.  However, they do state that 441 means "Armed in STAY mode" rather than STAY/INSTANT.  This makes sense, because in the log I sent you, I did not arm in instant mode, but in Night-Stay mode.

FYI,  the Envisalink TPI  references the 0x02 command received over the ECP bus in "Section 3.4: Partition Status Codes" on the ECP bus from the control board, rather than the 0xF7 command received over the ECP bus.  

From Dilbert66's code on github, you can see that the keypad update command 0xF7 not only has the night mode indicator, but also a separate instant mode indicator.  I'm sure that the F7 command is how the Honeywell 6150 Fixed English Keypad is able to display the "NIGHT" mode indicator when I use your integration to arm the panel in night mode.

I'll ask the folks at EyezOn why they don't decode this and provide appropriate indicators, and if they would update to include this capability.

`// Used to read bits on F7 message
#define BIT_MASK_BYTE1_BEEP 0x07
#define BIT_MASK_BYTE1_NIGHT 0x10

#define BIT_MASK_BYTE2_ARMED_HOME 0x80
#define BIT_MASK_BYTE2_LOW_BAT 0x40
#define BIT_MASK_BYTE2_ALARM_ZONE 0x20
#define BIT_MASK_BYTE2_READY 0x10
#define BIT_MASK_BYTE2_AC_LOSS 0x08//???
#define BIT_MASK_BYTE2_SYSTEM_FLAG 0x04
#define BIT_MASK_BYTE2_CHECK_FLAG 0x02
#define BIT_MASK_BYTE2_FIRE 0x01

#define BIT_MASK_BYTE3_INSTANT 0x80
#define BIT_MASK_BYTE3_PROGRAM 0x40
#define BIT_MASK_BYTE3_CHIME_MODE 0x20
#define BIT_MASK_BYTE3_BYPASS 0x10
#define BIT_MASK_BYTE3_AC_POWER 0x08
#define BIT_MASK_BYTE3_ARMED_AWAY 0x04
#define BIT_MASK_BYTE3_SYSTEM_ALARM 0x02
#define BIT_MASK_BYTE3_IN_ALARM 0x01

#define F7_MESSAGE_LENGTH 45`

**Comment by blfriedman** [comment]
FYI, I did contact EyezOn, and they simply stated

"We are not the manufacturer of the Envisalink. We have no control over the TPI interface.  You need to ask your HA developer to take it up with them.

Rather than ask you, I have sent a request for support directly to enquiries@envisacor.com to see if they could provide an updated firmware and TPI which includes the integration of the bitfields into the TPI which are included in the F7 status message sent on the ECP bus.

I'll let you know what I hear back.  It is surprising that this has not been a problem before - maybe night mode is used by many folks with the Vista-20 but knowing which mode is set programmatically is not that common.  However, I really would like to be able to utilize automations on night mode.

**Comment by ufodone** [comment]
> It is surprising that this has not been a problem before

I have a DSC system myself so don't know much about the Honeywell systems.  I think that the night mode for my system defaults to a zero entry delay - presumably because it is meant to be set when everyone is at home so if a fault does occur on an entry door then the alarm goes off immediately.  So it's possible other people with Honeywell systems have a similar setup.

**Comment by Dilbert66** [comment]
Night mode is indicated in the beep byte from the F7 cmd.  Lower 4 bits are the beep counts and beep types (ie exit delay and urgent).  Bit 5 (0x10) is the night mode indicator.  So this info should technically be available in Envisalinks beep field.  If not, then it means they are filtering out the upper 4 bits.  Put your system in night mode by sending <code>33 and have a look at the field after the exit delay completes.

**Comment by ufodone** [comment]
Thanks, @Dilbert66 .  That info is not included in the TPI documentation for the Envisalink.  It only says:

```
BEEP Field
This provides information to the virtual keypad on how to “beep”. 0 = OFF
1,2,3 = Beep this many times
4 = Continuous Fast Beep (trouble/urgency)
5 = Continuous Slow Beep (exit delay)
```

Looking at the debug logs @blfriedman provided, it looks like the EVL is setting the beep field to `03` on an initial update and then remains stable with a `10` value which could represent night mode.  So this BEEP field could be representing night mode. 

@blfriedman Could you capture some similar logs when arming your system in different modes so I can see what values the EVL is using for this?

**Comment by Dilbert66** [comment]
Yes, the 10 value is the night mode.  You will see the beep field start with a 5 indicating the exit beep, which then changes to 4 in the last few seconds of the exit delay, then will switch to 10 to indicate fully armed night mode.  EVL is displaying the exact data from the vista f7 beep field with no translation.  I suspect the TDI docs are a bit dated.  I'm using information from my own investigation and testing.

**Comment by Dilbert66** [comment]
I'll post some logs later if you wish when I get a chance.  That night mode flag in the beep field is the only one missing for that field.

**Comment by blfriedman** [comment]
Thanks to both you and Dilbert66 for following up on this.  It is his project I quoted above.

I have attached a debug log showing arming and disarming in STAY, AWAY, and NIGHT-STAY modes.

Not sure if I am looking at the right lines in the debug log file, but I see the following RX sequences for entering each arming mode.  I think it is especially interesting when entering NIGHT-STAY mode, it first shows STAY mode then switches to NIGHT-STAY, and the last byte in the log entry changes from 03 to 10.  If this is the byte 1 of the F7 message, then maybe that is it.

```
for STAY mode
RX < ^03,00$
RX < %02,0400000000000000$
RX < %00,01,8028,60,03,ARMED ***STAY***May Exit Now  60$

for AWAY mode
RX < ^03,00$
RX < %02,0500000000000000$
RX < %00,01,002C,60,02,ARMED ***AWAY***May Exit Now  60$

for NIGHT-STAY mode (note that it first shows ARMED ***STAY*** then transitions to ***NIGHT-STAY***)
RX < ^03,00$
RX < %02,0400000000000000$
RX < %00,01,8028,60,03,ARMED ***STAY***May Exit Now  60$
RX < ^00,00$
RX < %02,0600000000000000$
RX < %00,01,8028,60,10,***NIGHT-STAY***May Exit Now  60$
RX < %00,01,8028,59,10,***NIGHT-STAY***May Exit Now  59$
RX < %00,01,8028,58,10,***NIGHT-STAY***May Exit Now  58$
RX < %03,3441010020$
RX < %00,01,8028,57,10,***NIGHT-STAY***May Exit Now  57$
```

[arm_night-stay.log](https://github.com/user-attachments/files/17605081/arm_night-stay.log)

**Comment by ufodone** [comment]
Thanks for the details.  I'll try and get a fix out soon.  Having looked at the code a bit more, it looks like the night mode handling for DSC systems might also be not quite right.

**Comment by ufodone** [comment]
I've included a fix in the v0.9.4 release.  I don't have a Honeywell system myself but was able to simulate the scenario from your logs so I think it should be working properly now.

**Comment by blfriedman** [comment]
Thank you for getting this fix in there so quickly.  I tested it last night and it worked just as expected, I used the keypad to enable night mode on Honeywell/Vista and it was properly reflected in the status in Home Assistant.  Very helpful when I need to distinguish between normal "stay" mode and "night-stay" for automations.  Again, thank you and Dilbert66 for your work on the Vista system and your work on Envisalink.
